src/session-pool.ts

```diff
@@ -542,62 +548,67 @@ export class SessionPool extends EventEmitter implements SessionPoolInterface {
   }
 
   /**
-   * Attempts to create a session of a certain type.
+   * Attempts to create a single session of a certain type.
    *
    * @private
    *
    * @param {string} type The desired type to create.
    * @returns {Promise}
    */
-  async _createSession(type: types): Promise<void> {
-    const session = this.database.session();
-    const labels = this.options.labels!;
-
-    this._inventory.borrowed.add(session);
+  _createSession(type: types): Promise<void> {
+    const kind = type === types.ReadOnly ? 'reads' : 'writes';
+    const options = {[kind]: 1};
 
-    const createSession = async (): Promise<void> => {
-      await session.create({labels});
-
-      if (type === types.ReadWrite) {
-        try {
-          await this._prepareTransaction(session);
-        } catch (e) {
-          type = types.ReadOnly;
-        }
-      }
-    };
-
-    try {
-      await this._requests.add(createSession);
-    } catch (e) {
-      this._inventory.borrowed.delete(session);
-      throw e;
-    }
-
-    session.type = type;
-    session.lastUsed = Date.now();
-
-    this._inventory[type].push(session);
-    this._inventory.borrowed.delete(session);
+    return this._createSessions(options);
   }
 
   /**
-   * Attempts to create a session but emits any errors that occur.
+   * Batch creates sessions and prepares any necessary transactions.
    *
    * @private
    *
-   * @emits SessionPool#available
-   * @emits SessionPool#error
-   * @param {string} type The desired type to create.
+   * @param {object} [options] Config specifying how many sessions to create.
    * @returns {Promise}
    */
-  async _createSessionInBackground(type: types): Promise<void> {
+  async _createSessions({
+    reads = 0,
+    writes = 0,
+  }: CreateSessionsOptions): Promise<void> {
+    const count = reads + writes;
+    const labels = this.options.labels!;
+
+    this._pending += count;
+
+    let sessions: Session[] | null = null;
+
     try {
-      await this._createSession(type);
-      this.emit('available');
+      [sessions] = await this.database.batchCreateSessions({count, labels});
     } catch (e) {
-      this.emit('error', e);
+      this._pending -= count;
+      throw e;
     }
+
+    sessions.forEach(
+      async (session: Session): Promise<void> => {
+        let type = --writes > 0 ? types.ReadWrite : types.ReadOnly;
+
+        if (type === types.ReadWrite) {
+          try {
+            await this._prepareTransaction(session);
+          } catch (e) {
+            type = types.ReadOnly;
+          }
+        }
+
+        session.type = type;
+        session.lastUsed = Date.now();
+
+        this._pending -= 1;
+        this._inventory[type].push(session);
```

Comment:
Does Node.js have multiple gRPC channels? If there are multiple channels then the sessions are meant to be shuffled across the channels.

Reply:
We use a library called `grpc-gcp` that pools channels for us, so yes!

Reply:
Ah ok. I don't know if this works in the same way in Node.js. But in other languages, if a session is created through a given channel, then that session has an affinity to that channel. The problem when batchCreateSessions is introduced is that you allow a bunch of sessions to be created by a single channel, especially in the pool initialization phase, so that channel ends up being more utilized than other channels as the user makes requests.

This is why the code example in the spec is something like:
```
    // Spread session creation evenly over all available channels.
    int sessions_per_channel = min_sessions / num_channels;
    for (int i = 0; i < num_channels; i++) {
        sessions = batchCreateSessions(sessions_per_channel);
    }
```

Is there a way to get the number of channels you have available and spread the calls to batchCreateSessions over all the channels so that each channel gets its own portion of the sessions that are initialized?

Reply:
Hi @skuruppu @callmehiphop ,

The spec of grpc-gcp channel pool is to spin out new channels based on the number of concurrent streams on existing channels. So every time the number of open streams exceed the watermark value (configurable by client) for all the existing channels, a new channel will be created and added to the channel pool, until it reaches the upper bound pool size (also configurable by client). 

Can you help me better understand this batchCreateSessions function?

1. Is this a new Unary gRPC method (i.e. a single request with single response) or just a wrapper of existing `CreateSession` call? If it's a new unary rpc, I assume the response contains a list of created sessions from server. Then in this case, we would need to support batch BIND (binding session affinity to a channel). And I believe there are similar works being done in grpc-gcp-csharp to support batch BIND.

2. Are sessions expected to be evenly distributed across channels? If so, we may need to add this requirement to the spec of grpc-gcp and add this feature in grpc-gcp across all languages .

Reply:
> 1. Is this a new Unary gRPC method (i.e. a single request with single response) or just a wrapper of existing CreateSession call? 

It's a unary gRPC method so the response contains a list of created sessions from server as you said. So sessions currently don't have an affinity to the channel that it was created on?

> 2. Are sessions expected to be evenly distributed across channels?

Yes because if there are multiple channels, we want to utilize them as evenly as possible. In Java for example, if you call BatchCreateSessions on a particular channel, the returned sessions have an affinity to that channel. So we just need to call BatchCreateSessions using the logic I put in the code snippet above. Based on what you said in point 1, it doesn't work that way in Node.js currently without batch BIND.

I don't want to block this PR based on this discussion since it seems like a few other changes need to be made to support evenly distributing sessions across channels. So I'm happy to file an issue to do this work later.

Let me know if this works for you @WeiranFang.

Reply:
Resolving conversation since the changes that need to be made are out of scope for this PR.

Reply:
Thanks for the input! Then I believe we have same requirements (batch bind support & session balanced connection pool) for other languages as well. Please feel free to file an issue to grpc-gcp, also PR is welcomed :)